### ISSUE TITLE
tk.ValidationError returns HTTP 200

### ISSUE THREAD

**Issue by frafra** [ckan/ckan#6707]
**CKAN version**
master

**Describe the bug**
`tk.ValidationError` produces an error page, with HTTP code 200 instead of 400.

**Steps to reproduce**
Add invalid metadata for a dataset.

**Expected behavior**
The response should use an HTTP error code, and not an HTTP success code.

**Additional details**
https://github.com/ckan/ckanext-spatial/pull/272#issuecomment-1043211710

**Comment by TomeCirun** [comment]
@frafra How to add invalid metadata using the browser? 

Thanks

**Comment by frafra** [comment]
@TomeCirun just write something which is not accepted by CKAN, like not filling out a mandatory field.

**Comment by smotornyuk** [comment]
Good point. In most cases, edit/create pages use [code similar to this](https://github.com/ckan/ckan/blob/master/ckan/views/resource.py#L375-L380). 
Whenever validation error happens, extra variables `error/error_summary` added to template. But everything else works exactly as for normal GET request, i.e. page is rendered with 200 status code

As a quick workaround, instead of 
```
if ValidationError:
    return self.get(data, error)
```

You can use something like
```
if ValidationError:
    return self.get(data, error), 409
```

But if someone volunteers to work on it, it would be really cool to write something more explicit/reusable, if possible.

**Comment by amercader** [comment]
I think returning a 200 status error makes sense, you are generally rendering a page that lists some errors on your input data (as opposed to calling an API directly). 
@frafra do you think this a standard practice when inputing data via the browser?

**Comment by TomeCirun** [comment]
I found something that could be useful...

<img width="676" alt="image" src="https://user-images.githubusercontent.com/72216462/155277734-23bc7e54-6ef5-45db-aa0f-b0d2879a9049.png">

**Comment by frafra** [comment]
@amercader there is no standard practice when replying to a bad form request, but there are good practices, in my opinion. While everybody agrees that the correct HTTP status code should be used (4xx for handling errors, like validation, duplicates, authorization, and such) for JSON/XML REST API, there is less interest in form validation for some reason, so there is no common practice. Most advices in StackOverflow I found are in favor of using the correct HTTP status code for form validation too:
- https://stackoverflow.com/questions/7939137/what-http-status-code-should-be-used-for-wrong-input
- https://stackoverflow.com/questions/15580969/should-i-use-http-4xx-to-indicate-html-form-errors
- https://stackoverflow.com/questions/38846625/which-http-status-for-rendered-html-response-to-invalid-form-input

Django uses HTTP status code 200 as default just because of historical reasons, for example: https://code.djangoproject.com/ticket/22591.

**Comment by frafra** [comment]
Meeting notes 01/04/2022: 422 could be a good code; this issue is not a priority; we should warn user if we change HTTP status codes (potential issues with reverse proxies).
